lib/web_ui/dev/test_runner.dart

```diff
@@ -269,13 +269,14 @@ class TestCommand extends Command<bool> with ArgUtils {
       }
       // Run all unit-tests as a single batch.
       await _runTestBatch(unitTestFiles, concurrency: 10, expectFailure: false);
-      _checkExitCode();
+      await _checkExitCode();
     }
   }
 
-  void _checkExitCode() {
+  void _checkExitCode() async {
     if (io.exitCode != 0) {
       io.stderr.writeln('Process exited with exit code ${io.exitCode}.');
+      await cleanup(browser: browser);
```

Comment:
Instead of putting `cleanup` in multiple places, we should adopt the approach used by Flutter Tools, which is to throw a special exception (they use `ToolExit`, we can call our the same). Then we can have one place (probably the root `main()` function) that catches this exception and performs the clean-up.

Reply:
thanks this will be a very good improvement!

Reply:
done. this lead to a lot changes. there were other places I forgot to cleanup before exit.